Source/CarthageKit/BinariesCache.swift

```diff
@@ -153,6 +153,15 @@ final class BinaryProjectCache: AbstractBinariesCache {
                 let response = result.1
                 if let httpResponse = response as? HTTPURLResponse, !(200...299).contains(httpResponse.statusCode) {
                     return SignalProducer(error: CarthageError.httpError(statusCode: httpResponse.statusCode))
+                } else if Archive.hasTarExtension(fileURL: sourceURL) {
+                    let tempURL = destinationURL.deletingLastPathComponent().appendingPathComponent(sourceURL.lastPathComponent)
```

Comment:
Looks good, the only thing I would change is to do all the work in a temporary directory before moving it to the final location (to avoid concurrency issues):

```
private func rezipIfNeeded(sourceURL: URL, downloadURL: URL) -> SignalProducer<URL, CarthageError> {
        if Archive.hasTarExtension(fileURL: sourceURL) {
            // Untar and zip again
            let tempURL = downloadURL.deletingLastPathComponent().appendingPathComponent(sourceURL.lastPathComponent)
            return Files.moveFile(from: downloadURL, to: tempURL)
                .flatMap(.concat) { Archive.unarchive(archive: $0) }
                .flatMap(.concat) { directoryURL -> SignalProducer<URL, CarthageError> in
                    let zipArchiveFileName = "archive.zip"
                    let zipArchiveURL = directoryURL.appendingPathComponent(zipArchiveFileName)
                    return Archive.zip(paths: ["."], into: zipArchiveURL, workingDirectoryURL: directoryURL)
                }
        } else {
            return SignalProducer<URL, CarthageError>(value: downloadURL)
        }
    }
```

And call it as follows:

```
if let httpResponse = response as? HTTPURLResponse, !(200...299).contains(httpResponse.statusCode) {
                    return SignalProducer(error: CarthageError.httpError(statusCode: httpResponse.statusCode))
                } else {
                    return self.rezipIfNeeded(sourceURL: sourceURL, downloadURL: downloadURL)
                        .flatMap(.concat) { archiveURL -> SignalProducer<URL, CarthageError> in
                            return Files.moveFile(from: archiveURL, to: destinationURL)
                        }
                }
```

Reply:
Good point, I updated the PR to incorporate the suggestions